### PR TITLE
PersistentMap Improvements

### PR DESCRIPTION
**Pull request by eile**
* API for async writes: setQueueDepth()
* API for buffered reads: fetch()
* Support for builtin byteswapping
* templated get() for PODs
* Fix reading values >64k
* Performance and unit test

### CODE REVIEW

**Comment by eile** [comment]
@bilgili @biddisco FYI/for review

File: doc/RelNotes.md [hunk 1]
```diff
@@ -32,42 +24,21 @@ development and deployment of multi-threaded applications:
...
+* HDF4::registerLZF() function to register a LZF compression filter with
+  the HDF5 library
 * lunchbox::URI parsing support according to RFC3986
 * lunchbox::PersistentMap interface for persistent key-value stores
 * lunchbox::getHostname and lunchbox::getExecutablePath functions
 * lunchbox::PluginFactory, a mechanism to register and load plugins
-
-## Enhancements {#Enhancements}
-
 * Added timeout parameter for the Future::wait function
```
Comment (bilgili) [inline doc/RelNotes.md]: I think for classes like futures, random numbers, etc we should really re-consider either wrapping the boost ones or replace with boost ones.

**Comment by tribal-tec** [comment]
I don't understand why the hdf5 stuff needs to be in lunchbox now? where is it used?

File: doc/RelNotes.md (hunk 1 above)
Comment (eile) [inline doc/RelNotes.md]: To late. @biddisco is already working on returning a boost future from PersistentMap::get() 

:P

**Comment by eile** [comment]
The BBIC livre source. I had it in Brion first, but Livre does not depend on it.

File: doc/RelNotes.md (hunk 1 above)
Comment (biddisco) [inline doc/RelNotes.md]: >> To late. @biddisco<https://github.com/biddisco> is already working on returning a boost future from PersistentMap::get()

Correct and I agree that using boost:future would be a good move. You don’t want your own futures, boost futures and std::futures all mixed up. Future islands. etc etc



JB

File: lunchbox/hdf5/lzf/lzf.h [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+ * the above. If you wish to allow the use of your version of this file
+ * only under the terms of the GPL and not to allow others to use your
+ * version of this file under the BSD license, indicate your decision
+ * by deleting the provisions above and replace them with the notice
+ * and other provisions required by the GPL. If you do not delete the
+ * provisions above, a recipient may use your version of this file under
+ * either the BSD or the GPL.
+ */
+
+#ifndef LZF_H
```
Comment (bilgili) [inline lunchbox/hdf5/lzf/lzf.h]: I know this headers are not exposed but I had problems with packages having them. For instance Tuvok library has them too and conflicts with lunchbox on runtime. Can not we use lunchbox namespace for headers ?

**Comment by tribal-tec** [comment]
Hmpf, the hdf5 dependency here is a bummer to me, even it might be handy now to have it here. The bbic data source should depend on brion or be even in brion, but making lunchbox 'paying' for this is not the right thing to me.

**Comment by eile** [comment]
I see your point. Let me think about how to do it 'nice'. Makes also sense from the POV that Brion has an HDF5 lock already...

**Comment by bilgili** [comment]
And,  the future of BBIC is not yet clear !

**Comment by tribal-tec** [comment]
+1
Please update PR title

**Comment by eile** [comment]
Retest this please

**Comment by eile** [comment]
@tribal-tec: Updated with squashed work from the last weeks, pleas review again.

**Comment by eile** [comment]
Retest this please

File: tests/persistentMap.cpp [hunk 3]
```diff
@@ -117,22 +266,35 @@ void testLevelDBFailures()
...
 {
+    const bool perfTest LB_UNUSED
+        = std::string( argv[0] ).find( "perf_" ) != std::string::npos;
+    if( perfTest )
+        std::cout << "  async,    read,   write,   fetch" << std::endl;
+
     try
     {
-#ifdef LUNCHBOX_USE_LEVELDB
+#ifdef __LUNCHBOX_USE_LEVELDB
```
Comment (tribal-tec) [inline tests/persistentMap.cpp]: why leading underscores? does not match the rest of the file

File: lunchbox/persistentMap.h [hunk 4]
```diff
@@ -145,11 +185,15 @@ class PersistentMap : public boost::noncopyable
     /** Flush outstanding operations to the backend storage. @version 1.10 */
     LUNCHBOX_API bool flush();
 
+    /** Enable or disable endianness conversion on reads. @version 1.9.2 */
+    LUNCHBOX_API void setByteswap( const bool swap );
```
Comment (tribal-tec) [inline lunchbox/persistentMap.h]: not used?

File: tests/persistentMap.cpp (hunk 3 above)
Comment (eile) [inline tests/persistentMap.cpp]: Sorry, slipped through from John

File: lunchbox/persistentMap.h (hunk 4 above)
Comment (eile) [inline lunchbox/persistentMap.h]: Used by brion::CompartmentReportMap

Comment (tribal-tec) [inline lunchbox/persistentMap.h]: missing unit test then :)

File: tests/persistentMap.cpp (hunk 3 above)
Comment (biddisco) [inline tests/persistentMap.cpp]: Sorry. It was indeed me. I changed the ifdef to enable syntax highlightin in eclipse and then forgot to reset it. You may find others like that

JB

File: lunchbox/persistentMap.h (hunk 4 above)
Comment (eile) [inline lunchbox/persistentMap.h]: :P

done.

File: tests/persistentMap.cpp (hunk 3 above)
Comment (eile) [inline tests/persistentMap.cpp]: done

Comment (tribal-tec) [inline tests/persistentMap.cpp]: @eile still there?

**Comment by tribal-tec** [comment]
Retest this please

**Pull request merged**